kommandir/bin/adept_openstack.py

```diff
@@ -0,0 +1,720 @@
+        self.floating_network_id = gw_info['network_id']
+        logging.debug("Router %s network id: %s",
+                      routers[0]['name'],
+                      self.floating_network_id)
+
+    def done(self):
+        """Return true when server ID has a floating IP address"""
+        try:
+            self.os_rest.server_ip(uuid=self.server_id)
+            return True
+        except (ValueError, KeyError, IndexError):
+            logging.info("Assignment failed, retrying")
+            time.sleep(0.5)  # don't busy-wait
+            return False
+
+    def action(self):
+        """Assign an existing or new floating ip to server, return ip"""
+        floating_ip = self.os_rest.floating_ip()
+        if not floating_ip:
+            logging.info("Creating new floating IP address on network %s",
+                         self.floating_network_id)
+            floatingip = dict(floating_network_id=self.floating_network_id)
+
+            self.os_rest.service_request('network', '/v2.0/floatingips',
+                                         "floatingip", method='post',
+                                         post_json=dict(floatingip=floatingip))
+            floating_ip = self.os_rest.response_json['floating_ip_address']
+        else:
+            logging.info("Found disused ip %s", floating_ip)
+
+        logging.info("Assigning floating IP to server")
+        addfloatingip = dict(address=floating_ip)
+        self.os_rest.compute_request('/servers/%s/action' % self.server_id,
+                                     unwrap=None, method='post',
+                                     post_json=dict(addFloatingIp=addfloatingip))
+        return floating_ip
+
+
+def create(name, pub_key_files):
+    """
+    Create a new VM with name and authorized_keys containing pub_key_files.
+
+    :param name: Name of the VM to create
+    :param pub_key_files: List of ssh public key files to read
+    :param basename: Not used
+    """
+
+    pubkeys = []
+    for pub_key_file in pub_key_files:
+        logging.debug("Loading public key file: %s", pub_key_file)
+        with open(pub_key_file, 'rb') as key_file:
+            pubkeys.append(key_file.read().strip())
+            if 'PRIVATE KEY' in pubkeys[-1]:
+                raise ValueError("File %s appears to be a private, not a public, key"
+                                 % pub_key_file)
+    logging.info("Deleting any partially created VM %s", name)
+    TimeoutDeleted(name)()
+
+    try:
+        server_id = TimeoutCreate(name, pubkeys)()
+        logging.info("Creation successful, attempting to assign floating ip to VM %s", name)
+        floating_ip = TimeoutAssignFloatingIP(server_id)()
+        logging.info("Floating IP assignment successful, ip %s", floating_ip)
+    except:
+        # Fire and forget
+        os_rest = OpenstackREST()
+        os_rest.server_delete(name)
+        raise
+    sys.stdout.write(OUTPUT_FORMAT.format(name=name, floating_ip=floating_ip))
+    return 0
```

Comment:
It takes some effort to figure out that `0` here is a UNIX exit code which will eventually be passed to sys.exit(). Since none of `create`, `discover`, or `delete` ever return anything other than 0, why bother? Why not just fall through without an explicit return value? That will be less confusing to the reader.